### PR TITLE
Allow rapids_test to be used without CUDAToolkit

### LINKED ISSUE #478: [FEA] Allow `rapids_test` to be used without CUDAToolkit
**Is your feature request related to a problem? Please describe.**

`rapids_test_init()` calls `find_package(CUDAToolKit REQUIRED)` if it does not find a `generate_resource_spec` executable at exactly `<build-dir>/rapids-cmake/generate_resource_spec`

Obviously this doesn't work so well on cpu-only builds.

**Describe the solution you'd like**

Provide a way of disabling this check, or provide an option `USE_GPU OFF`. Ideally we could simply do
```cmake
include(path/to/rapids-test.cmake)

rapids_test_init(ENABLE_GPU OFF)
```

**Describe alternatives you've considered**

Currently, we work around this by spoofing the file like so
```cmake
include(${rapids-cmake-dir}/test/detail/default_names.cmake)

set(patch_file "${CMAKE_CURRENT_SOURCE_DIR}/cmake/patches/generate_resource_spec.diff")
set(eval_file ${rapids-cmake-dir}/test/detail/generate_resource_spec.cpp)
set(eval_exe ${CMAKE_CURRENT_BINARY_DIR}/rapids-cmake/${rapids_test_generate_exe_name})

message("-- Patching ${eval_file}")
find_program(PATCH NAMES patch REQUIRED)
execute_process(
  COMMAND ${PATCH} --unified ${eval_file} --input=${patch_file} --ignore-whitespace --quiet
  TIMEOUT 15
  COMMAND_ERROR_IS_FATAL ANY
)

# Compile the (now patched) resource spec generator, and place the result where rapids
# expects to find it. try_compile() is a bit of a misnomer here -- this compilation MUST
# succeed, as rapids_test_init() below will fail if it doesn't
try_compile(
  SUCCEEDED
  SOURCES ${eval_file}
  COPY_FILE ${eval_exe}
  OUTPUT_VARIABLE compile_ret
)

if(NOT SUCCEEDED)
  message(FATAL_ERROR "Failed to compile patched resource spec generator for rapids_cmake_test(): ${compile_ret}" )
endif()
```
where `generate_resource_spec.diff` is
```diff
--- ./generate_resource_spec.cpp
+++ ./generate_resource_spec.cpp
@@ -14,7 +14,6 @@
  * limitations under the License.
  */

-#include <cuda_runtime_api.h>

 #include <iostream>
 #include <string>
@@ -27,7 +26,6 @@

 struct gpu {
   gpu(int i) : id{i} {};
-  gpu(int i, const cudaDeviceProp& prop) : id{i}, memory{prop.totalGlobalMem}, slots{100} {}
   int id        = 0;
   size_t memory = 0;
   int slots     = 0;
@@ -47,17 +45,8 @@
 int main()
 {
   std::vector<gpu> gpus;
-  int nDevices = 0;
-  cudaGetDeviceCount(&nDevices);
-  if (nDevices == 0) {
-    gpus.emplace_back(0);
-  } else {
-    for (int i = 0; i < nDevices; ++i) {
-      cudaDeviceProp prop;
-      cudaGetDeviceProperties(&prop, i);
-      gpus.emplace_back(i, prop);
-    }
-  }
+
+  gpus.emplace_back(0);

   version v{1, 0};
   std::cout << "{\n";
```

### PR DESCRIPTION
**Pull request by robertmaynard**
## Description
Fixes #478

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/rapids-cmake/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.
- [x] The `cmake-format.json` is up to date with these changes.

### CODE REVIEW

File: rapids-cmake/test/detail/generate_resource_spec.cpp [hunk 1]
```diff
@@ -48,16 +50,21 @@ int main()
...
     for (int i = 0; i < nDevices; ++i) {
       cudaDeviceProp prop;
       cudaGetDeviceProperties(&prop, i);
-      gpus.emplace_back(i, prop);
+      gpus.emplace_back(i, prop.totalGlobalMem);
     }
   }
+#else
+  gpus.emplace_back(0);
+#endif()
```
Comment (Jacobfaib) [inline rapids-cmake/test/detail/generate_resource_spec.cpp:67]:
```suggestion
#endif
```

File: rapids-cmake/test/generate_resource_spec.cmake [hunk 2]
```diff
@@ -67,10 +67,12 @@ function(rapids_test_generate_resource_spec DESTINATION filepath)
...
 
   if(NOT EXISTS "${eval_exe}")
-    find_package(CUDAToolkit REQUIRED)
+    find_package(CUDAToolkit)
     file(MAKE_DIRECTORY "${PROJECT_BINARY_DIR}/rapids-cmake/")
 
-    set(compile_options "-I${CUDAToolkit_INCLUDE_DIRS}")
+    if(CUDAToolkit_FOUND)
+      set(compile_options "-I${CUDAToolkit_INCLUDE_DIRS}" "-DHAVE_CUDA")
+    endif()
```
Comment (Jacobfaib) [inline rapids-cmake/test/generate_resource_spec.cmake:75]: I get many warnings building with Apple `clang` (it stupidly still defaults to C++98):
```
_deps/rapids-cmake-src/rapids-cmake/test/detail/generate_resource_spec.cpp:69:11: error: no matching constructor for initialization of 'version'
  version v{1, 0};
          ^~~~~~~
_deps/rapids-cmake-src/rapids-cmake/test/detail/generate_resource_spec.cpp:27:13: warning: default member initializer for non-static data member is a C++11 extension [-Wc++11-extensions]
  int minor = 0;
            ^
_deps/rapids-cmake-src/rapids-cmake/test/detail/generate_resource_spec.cpp:31:22: error: expected member name or ';' after declaration specifiers
  gpu(int i) : id{i} {};
                     ^
_deps/rapids-cmake-src/rapids-cmake/test/detail/generate_resource_spec.cpp:32:33: error: expected member name or ';' after declaration specifiers
  gpu(int i, size_t mem) : id{i}, memory{mem}, slots{100} {}
                                ^
_deps/rapids-cmake-src/rapids-cmake/test/detail/generate_resource_spec.cpp:34:17: warning: default member initializer for non-static data member is a C++11 extension [-Wc++11-extensions]
  size_t memory = 0;
                ^
_deps/rapids-cmake-src/rapids-cmake/test/detail/generate_resource_spec.cpp:35:17: warning: default member initializer for non-static data member is a C++11 extension [-Wc++11-extensions]
  int slots     = 0;
                ^
_deps/rapids-cmake-src/rapids-cmake/test/detail/generate_resource_spec.cpp:31:18: error: expected '('
  gpu(int i) : id{i} {};
                 ^
```
This is fixed by the following
```suggestion
    else()
      set(compile_options "-std=c++14")
    endif()
```
Perhaps better to add this flag also for the GPU build.

**Review by Jacobfaib — CHANGES_REQUESTED** [review]
A few minor issues. Also another more minor nit, the configure steps here are quite chatty:
```
-- Could not find nvcc, please set CUDAToolkit_ROOT.
'/opt/homebrew/opt/ccache/libexec/c++' '/path/to/_deps/rapids-cmake-src/rapids-cmake/test/detail/generate_resource_spec.cpp' '-o' '/path/to/rapids-cmake/generate_ctest_json'
```
Is it possible to silence these unless there is a problem (or extra verbosity is requested)? The first message implies there is a problem (which there isn't), while the second is not useful for the user to see unless there is an issue.

File: rapids-cmake/test/generate_resource_spec.cmake (hunk 2 above)
Comment (robertmaynard) [reply rapids-cmake/test/generate_resource_spec.cmake:75]: Corrected these errors by making sure the source is c++98 conformant. That way we can continue to support compilers that don't handle `-std=` flags

**Comment by robertmaynard** [comment]
> A few minor issues. Also another more minor nit, the configure steps here are quite chatty:
> 
> ```
> -- Could not find nvcc, please set CUDAToolkit_ROOT.
> '/opt/homebrew/opt/ccache/libexec/c++' '/path/to/_deps/rapids-cmake-src/rapids-cmake/test/detail/generate_resource_spec.cpp' '-o' '/path/to/rapids-cmake/generate_ctest_json'
> ```
> 
> Is it possible to silence these unless there is a problem (or extra verbosity is requested)? The first message implies there is a problem (which there isn't), while the second is not useful for the user to see unless there is an issue.

Can you try the updated version again, please? All your issues should be resolved now

**Comment by Jacobfaib** [comment]
> Can you try the updated version again, please? All your issues should be resolved now

OK tried latest 07e58cd46a636ca9b5632d2d9194a7ff0dc80794, and it seems that the build is still outputting
```sh
'/opt/homebrew/opt/ccache/libexec/c++' '/path/to/_deps/rapids-cmake-src/rapids-cmake/test/detail/generate_resource_spec.cpp' '-o' '/path/to/rapids-cmake/generate_ctest_json'
```
From what I can tell this is due to the [compile command](https://github.com/rapidsai/rapids-cmake/pull/480/files#diff-92f51ddffa5e81afb760f384afc6f2249d2b75ed468d9f4fbaa77bc8560a8ed1R82-R84) having the `COMMAND_ECHO STDOUT` option set. Perhaps this can be directed into a variable instead?

Also, I'm sure CI will flag this momentarily, but...
```
/path/to/_deps/rapids-cmake-src/rapids-cmake/test/detail/generate_resource_spec.cpp:43:76: error: no member named 'minor' in 'version'
  buffer << "\"version\": {\"major\": " << v.major << ", \"minor\": " << v.minor << "}";
                                                                         ~ ^
```

**Comment by robertmaynard** [comment]
> > Can you try the updated version again, please? All your issues should be resolved now
> 
> OK tried latest [07e58cd](https://github.com/rapidsai/rapids-cmake/commit/07e58cd46a636ca9b5632d2d9194a7ff0dc80794), and it seems that the build is still outputting the following:
> 
> ```shell
> '/opt/homebrew/opt/ccache/libexec/c++' '/path/to/_deps/rapids-cmake-src/rapids-cmake/test/detail/generate_resource_spec.cpp' '-o' '/path/to/rapids-cmake/generate_ctest_json'
> ```
> 
> From what I can tell this is due to the [compile command](https://github.com/rapidsai/rapids-cmake/pull/480/files#diff-92f51ddffa5e81afb760f384afc6f2249d2b75ed468d9f4fbaa77bc8560a8ed1R82-R84) having the `COMMAND_ECHO STDOUT` option set. Perhaps this can be directed into a variable instead?
> 
> Also, I'm sure CI will flag this momentarily, but...
> 
> ```
> /path/to/_deps/rapids-cmake-src/rapids-cmake/test/detail/generate_resource_spec.cpp:43:76: error: no member named 'minor' in 'version'
>   buffer << "\"version\": {\"major\": " << v.major << ", \"minor\": " << v.minor << "}";
>                                                                          ~ ^
> ```

Yeah I missed the `COMMAND_ECHO`, thanks. Removed that and only dump the compile info when we fail.

**Comment by Jacobfaib** [comment]
OK, with a quick fixup to fix that compile error [above](https://github.com/rapidsai/rapids-cmake/pull/480#issuecomment-1791204730), the latest LGTM.

**Review by Jacobfaib — APPROVED** [review]

**Pull request merged**